File: database/database-vertx/src/main/java/ru/tinkoff/kora/database/vertx/VertxConnectionFactory.java
package ru.tinkoff.kora.database.vertx;

import io.vertx.sqlclient.Pool;
import io.vertx.sqlclient.SqlConnection;
import reactor.core.publisher.Mono;
import ru.tinkoff.kora.database.common.telemetry.DataBaseTelemetry;

import java.util.concurrent.CompletionStage;
import java.util.function.Function;

public interface VertxConnectionFactory {
    SqlConnection currentConnection();

    CompletionStage<SqlConnection> newConnection();

    Pool pool();

    DataBaseTelemetry telemetry();

    <T> Mono<T> withConnection(Function<SqlConnection, Mono<T>> callback);

    <T> Mono<T> inTx(Function<SqlConnection, Mono<T>> callback);
}

File: database/database-vertx-coroutines/src/main/kotlin/ru/tinkoff/kora/database/vertx/coroutines/VertxConnectionFactory.kt
package ru.tinkoff.kora.database.vertx.coroutines

import io.vertx.sqlclient.Pool
import io.vertx.sqlclient.SqlConnection
import kotlinx.coroutines.ExperimentalCoroutinesApi
import ru.tinkoff.kora.database.common.telemetry.DataBaseTelemetry

@ExperimentalCoroutinesApi
interface VertxConnectionFactory {

    suspend fun currentConnection(): SqlConnection?

    suspend fun newConnection(): SqlConnection

    fun pool(): Pool

    fun telemetry(): DataBaseTelemetry

    suspend fun <T: Any?> withConnection(callback: suspend (SqlConnection) -> T): T

    suspend fun <T: Any?> inTx(callback: suspend (SqlConnection) -> T): T
}
